mm2src/coins/nft/storage/sql_storage.rs

```diff
@@ -1285,28 +1346,131 @@ impl NftTransferHistoryStorageOps for AsyncMutexGuard<'_, AsyncConnection> {
+        return Err(AsyncConnError::Internal(InternalError(
+            "Primary key duplication occurred in old nft tx history table".to_string(),
+        )));
+    }
+
+    // Start a transaction to ensure all operations are atomic
+    let sql_tx = conn.transaction()?;
+
+    // Create the temporary table with the new schema
+    let temp_table_name = SafeTableName::new(format!("{}_temp", history_table.inner()).as_str())?;
+    sql_tx.execute(&create_transfer_history_table_sql_custom_name(&temp_table_name)?, [])?;
+
+    // I don't think we need to batch the data copy process here.
+    // It's unlikely that the table will grow to 1 million+ rows (as an example).
+    let copy_data_sql = format!(
+        "INSERT INTO {} SELECT * FROM {};",
+        temp_table_name.inner(),
+        history_table.inner()
+    );
+    sql_tx.execute(&copy_data_sql, [])?;
+
+    let drop_old_table_sql = format!("DROP TABLE IF EXISTS {};", history_table.inner());
+    sql_tx.execute(&drop_old_table_sql, [])?;
+
+    let rename_table_sql = format!(
+        "ALTER TABLE {} RENAME TO {};",
+        temp_table_name.inner(),
+        history_table.inner()
+    );
+    sql_tx.execute(&rename_table_sql, [])?;
+
+    sql_tx.execute(&update_schema_version_sql(&schema_table), [
+        history_table.inner().to_string(),
+        CURRENT_SCHEMA_VERSION_TX_HISTORY.to_string(),
+    ])?;
+
+    sql_tx.commit()?;
+
+    Ok(())
+}
+
+/// Query to check for duplicates based on the primary key columns from tx history table version 2
+fn has_primary_key_duplication(conn: &Connection, safe_table_name: &SafeTableName) -> Result<bool, SqlError> {
+    let query = format!(
+        "SELECT EXISTS (
+            SELECT 1
+            FROM {}
+            GROUP BY transaction_hash, log_index, token_id
+            HAVING COUNT(*) > 1
+        );",
+        safe_table_name.inner()
+    );
+    // return true if duplicates exist, false otherwise
+    conn.query_row(&query, [], |row| row.get::<_, i32>(0))
+        .map(|exists| exists == 1)
+}
+
+#[async_trait]
+impl NftMigrationOps for AsyncMutexGuard<'_, AsyncConnection> {
+    type Error = AsyncConnError;
+
+    async fn migrate_tx_history_if_needed(&self, chain: &Chain) -> MmResult<(), Self::Error> {
+        let history_table = chain.transfer_history_table_name()?;
+        let schema_table = schema_versions_table_name()?;
+        self.call(move |conn| {
+            let schema_table_exists =
+                query_single_row(conn, CHECK_TABLE_EXISTS_SQL, [schema_table.inner()], string_from_row)?;
+
+            let version = if schema_table_exists.is_some() {
+                get_schema_version_stmt(conn)?
+                    .query_row([history_table.inner()], |row| row.get(0))
+                    .unwrap_or(0)
+            } else {
+                conn.execute(&create_schema_versions_sql()?, []).map(|_| ())?;
+                0
+            };
+
+            if version < CURRENT_SCHEMA_VERSION_TX_HISTORY {
```

Comment:
Same here

Reply:
Also did for sql https://github.com/KomodoPlatform/komodo-defi-framework/pull/2209/commits/7636464c657ca4cae5c494b910ea42fd83781d35

Reply:
Same comment https://github.com/KomodoPlatform/komodo-defi-framework/pull/2209#discussion_r1826062717

Reply:
> Same comment [#2209 (comment)](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2209#discussion_r1826062717)

for sql I suggest to do the same as for wasm https://github.com/KomodoPlatform/komodo-defi-framework/pull/2209#discussion_r1826954977